lib/realtime/vehicle.ex

```diff
@@ -63,19 +87,136 @@ defmodule Realtime.Vehicle do
     }
   """
   @spec decode(term()) :: t()
-  def decode(%{} = json) do
+  def decode(%{} = json, opts \\ []) do
+    opts = Keyword.merge(@default_opts, opts)
+    stops_on_route_fn = Keyword.get(opts, :stops_on_route_fn)
+    timepoints_on_route_fn = Keyword.get(opts, :timepoints_on_route_fn)
+
+    id = json["id"]
+    label = json["vehicle"]["vehicle"]["label"]
+    timestamp = json["vehicle"]["timestamp"]
+    direction_id = json["vehicle"]["trip"]["direction_id"]
+    route_id = json["vehicle"]["trip"]["route_id"]
+    trip_id = json["vehicle"]["trip"]["trip_id"]
+    current_status = decode_current_status(json["vehicle"]["current_status"])
+    stop_id = json["vehicle"]["stop_id"]
+
+    stops_on_route =
+      try do
+        stops_on_route_fn.(route_id)
+      catch
+        # Handle Gtfs server timeouts gracefully
+        :exit, _ ->
+          []
+      end
+
+    timepoints_on_route =
+      try do
+        timepoints_on_route_fn.(route_id)
+      catch
+        # Handle Gtfs server timeouts gracefully
+        :exit, _ ->
+          []
+      end
+
+    ordered_stops = if direction_id == 0, do: Enum.reverse(stops_on_route), else: stops_on_route
+
+    {percent_of_the_way_to_next_timepoint, next_timepoint} =
+      percent_of_the_way_to_next_timepoint(ordered_stops, timepoints_on_route, stop_id)
+
     %__MODULE__{
-      id: json["id"],
-      label: json["vehicle"]["vehicle"]["label"],
-      timestamp: json["vehicle"]["timestamp"],
-      direction_id: json["vehicle"]["trip"]["direction_id"],
-      route_id: json["vehicle"]["trip"]["route_id"],
-      trip_id: json["vehicle"]["trip"]["trip_id"],
-      current_status: decode_current_status(json["vehicle"]["current_status"]),
-      stop_id: json["vehicle"]["stop_id"]
+      id: id,
+      label: label,
+      timestamp: timestamp,
+      direction_id: direction_id,
+      route_id: route_id,
+      trip_id: trip_id,
+      current_status: current_status,
+      stop_id: stop_id,
+      status: %{
+        stop_granularity: %{
+          current_status: current_status,
+          stop_id: stop_id
+        },
+        timepoint_granularity: %{
+          current_status: timepoint_status(current_status, percent_of_the_way_to_next_timepoint),
+          timepoint_id: next_timepoint.id,
+          percent_of_the_way: percent_of_the_way_to_next_timepoint
+        }
+      }
+    }
+  end
+
+  @spec percent_of_the_way_to_next_timepoint([Stop.id()], [Timepoint.t()], Stop.id()) :: {
+          non_neg_integer(),
+          Timepoint.t()
+        }
+
+  def percent_of_the_way_to_next_timepoint(stop_ids, _timepoints, _stop_id)
+      when is_list(stop_ids) and length(stop_ids) == 0,
```

Comment:
Can just be `percent_of_the_way_to_next_timepoint([] = _stop_ids, _timepoints, _stop_id)` instead of the `when`.